### PR TITLE
PyBand: Transaction class, add test cases

### PR DESCRIPTION
**Pull request by evilpeach**
## Implementation details
- Transaction class as new spec
  - `with_messages`, `with_memo`, `with_account_num`, `with_sequence`, `with_fee`, `with_gas`
  - `get_sign_data`
  - `get_tx_data`
- Added Transaction's test cases
- Move mode to `send_tx` on Client

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [x] The pull request includes a description of the implementation/work done in detail.
- [x] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: helpers/pyband/pyband/transaction.py [hunk 1]
```diff
@@ -1,84 +1,99 @@
...
-            separators=(",", ":"),
-            sort_keys=True,
-        )
-        message_bytes = message_str.encode("utf-8")
+    def with_messages(self, *msgs: List[Msg]):
+        self.msgs.extend(msgs)
+        return self
 
-        signature = self.privkey.sign(message_bytes)
+    def with_account_num(self, v):
```
Comment (GriphookETH) [inline helpers/pyband/pyband/transaction.py:23]: types

File: helpers/pyband/pyband/transaction.py [hunk 2]
```diff
@@ -1,84 +1,99 @@
...
-                "ask_count": str(ask_count),
-                "min_count": str(min_count),
-                "client_id": client_id,
-                "sender": self.privkey.to_pubkey().to_address().to_acc_bech32(),
-            },
-        }
-        self.msgs.append(msg)
+    def __init__(self):
+        self.msgs: List[Msg] = []
+        self.account_num: int = None
```
Comment (GriphookETH) [inline helpers/pyband/pyband/transaction.py:12]: : int = None?

File: helpers/pyband/pyband/client.py [hunk 3]
```diff
@@ -30,8 +30,10 @@ def _get(self, path, **kwargs):
     def _get_result(self, path, **kwargs):
         return self._get(path, **kwargs)["result"]
 
-    def send_tx(self, data: dict) -> dict:
-        return requests.post(self.rpc_url + "/txs", json=data).json()
+    def send_tx(self, data: dict, mode="sync") -> dict:
```
Comment (GriphookETH) [inline helpers/pyband/pyband/client.py:33]: should we return something more specific that a dict

**Review by GriphookETH — CHANGES_REQUESTED** [review]

File: helpers/pyband/pyband/client.py (hunk 3 above)
Comment (sorawit) [inline helpers/pyband/pyband/client.py:33]: should return type be dict?

File: helpers/pyband/pyband/transaction.py (hunk 2 above)
Comment (sorawit) [inline helpers/pyband/pyband/transaction.py]: `: int = None`?

File: helpers/pyband/pyband/transaction.py (hunk 1 above)
Comment (sorawit) [inline helpers/pyband/pyband/transaction.py]: Add types

File: helpers/pyband/pyband/transaction.py [hunk 4]
```diff
@@ -1,84 +1,99 @@
...
+
+    def with_gas(self, v):
+        self.gas = v
+        return self
+
+    def with_memo(self, v):
+        self.memo = v
+        return self
+
+    def get_data_for_sign(self) -> bytes:
```
Comment (sorawit) [inline helpers/pyband/pyband/transaction.py]: def `get_sign_data(self)`

File: helpers/pyband/pyband/transaction.py [hunk 5]
```diff
@@ -1,84 +1,99 @@
...
+            },
+            "memo": self.memo,
+            "sequence": str(self.sequence),
+            "msgs": list(map(lambda x: x.as_json(), self.msgs)),
+        }
+
+        message_str = json.dumps(message_json, separators=(",", ":"), sort_keys=True)
+        return message_str.encode("utf-8")
+
+    def get_data_for_send(self, signature: bytes, pubkey: PublicKey) -> Dict:
```
Comment (sorawit) [inline helpers/pyband/pyband/transaction.py]: `def get_tx_data(self)`

File: helpers/pyband/pyband/transaction.py [hunk 6]
```diff
@@ -1,84 +1,99 @@
...
+        return {
+            "msg": list(map(lambda x: x.as_json(), self.msgs)),
```
Comment (sorawit) [inline helpers/pyband/pyband/transaction.py]: [x.as_json() for x in self.msgs] is more pythonic i think

File: helpers/pyband/pyband/transaction.py [hunk 7]
```diff
@@ -1,84 +1,99 @@
...
-            "sequence": str(sequence),
-            "msgs": self.msgs,
+            "account_number": str(self.account_num),
+            "fee": {
+                "gas": str(self.gas),
+                "amount": [{"amount": str(self.fee), "denom": "uband"}],
+            },
+            "memo": self.memo,
+            "sequence": str(self.sequence),
+            "msgs": list(map(lambda x: x.as_json(), self.msgs)),
```
Comment (sorawit) [inline helpers/pyband/pyband/transaction.py]: [x.as_json() for x in self.msgs] is more pythonic i think

**Review by sorawit — CHANGES_REQUESTED** [review]

File: helpers/pyband/pyband/client.py (hunk 3 above)
Comment (evilpeach) [reply helpers/pyband/pyband/client.py:33]: I'll make another PR for Client class spec.

**Review by taobun — APPROVED** [review]

File: helpers/pyband/pyband/transaction.py [hunk 8]
```diff
@@ -1,84 +1,103 @@
...
+            },
+            "memo": self.memo,
+            "sequence": str(self.sequence),
+            "msgs": [x.as_json() for x in self.msgs],
+        }
+
+        message_str = json.dumps(message_json, separators=(",", ":"), sort_keys=True)
+        return message_str.encode("utf-8")
+
+    def get_tx_data(self, signature: bytes, pubkey: PublicKey) -> Dict:
```
Comment (sorawit) [inline helpers/pyband/pyband/transaction.py]: should we use `dict`?

File: helpers/pyband/pyband/transaction.py [hunk 9]
```diff
@@ -1,84 +1,103 @@
...
+        self.memo: str = ""
 
-    def _sign(self, account_num: int, sequence: int, gas: int, fee: int, memo: str) -> str:
-        message_str = json.dumps(
-            self._get_sign_message(account_num, sequence, gas, fee, memo),
-            separators=(",", ":"),
-            sort_keys=True,
-        )
-        message_bytes = message_str.encode("utf-8")
+    def with_messages(self, *msgs: Msg):
```
Comment (sorawit) [inline helpers/pyband/pyband/transaction.py]: Add return type?

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**